### PR TITLE
Update acceptance tests to Selenium 3

### PR DESCRIPTION
**Pull request by danxuliu**
This should be rebased on master once #25983 is merged.

The acceptance tests used the last Selenium 2 Docker container available, which provides a rather old Firefox version (Firefox 47). Nevertheless, despite some rendering issues, most things still worked as expected due to the JavaScript files being built with support for older browsers. However, now that [support for Internet Explorer 11 and older browsers will be dropped](https://github.com/nextcloud/server/pull/25968) things could start to fail, so a newer browser (and thus a newer Selenium version) should be used in the acceptance tests.

[Selenium has been standardized by the W3C](https://w3c.github.io/webdriver/), and the protocol to communicate between the Selenium server and the browser has changed due to that. Firefox >= 48 only supports the new W3C protocol, but the Selenium driver for Mink does not support it yet (see issue 293 in https://github.com/minkphp/MinkSelenium2Driver/issues, not linking to prevent polluting the thread with back references).

The old protocol can still be used in recent Chromium/Chrome versions by explicitly forcing it, so for the time being the acceptance tests will need to be run on Chrome instead (although Firefox provides some interesting features like the fake streams that would be needed to test calls in Talk, so they should be moved again to Firefox once possible).

Finally, the default shm size of Docker is 64 MiB. This does not seem enough to run newer Chromium releases and causes the browser to randomly crash during the tests (`unknown error: session deleted because of page crash` is shown in the logs). Due to this [`disable-dev-shm-usage` needs to be used so Chrome writes shared memory files into _/tmp_ instead of _/dev/shm_](https://developers.google.com/web/tools/puppeteer/troubleshooting#tips) (the default shm size of Docker could have been increased instead using `docker run --shm-size...`, but that seems to be problematic when the container is run in current Drone releases).

~~**Pending:** shm size needs to be increased when run in Drone too. In theory [it should be enough to use `shm_size`](https://stackoverflow.com/questions/41546310/setting-up-drone-agents-to-have-custom-dev-shm/41640458#41640458), but [it seems that parameter no longer works](https://discourse.drone.io/t/shm-size-setting-not-respected-in-drone-1-0/4921) (although I have not actually tested it, I just hoped that for some reason our Drone had a high enough default shm size, but based on the `unknown error: session deleted because of page crash` errors in the logs it seems it does not :-( )~~

### CODE REVIEW

**Comment by skjnldsv** [comment]
@danxuliu promising results! :)

**Review by skjnldsv — APPROVED** [review]

File: tests/acceptance/features/bootstrap/FilesAppSharingContext.php [hunk 1]
```diff
@@ -427,7 +428,7 @@ public function iProtectTheSharedLinkWithThePassword($password) {
 		$shareLinkMenuTriggerElement = $this->actor->find(self::shareLinkMenuTrigger(), 2);
 		$this->actor->find(self::passwordProtectCheckbox($shareLinkMenuTriggerElement), 2)->click();
 
-		$this->actor->find(self::passwordProtectField($shareLinkMenuTriggerElement), 2)->setValue($password . "\r");
+		$this->actor->find(self::passwordProtectField($shareLinkMenuTriggerElement), 2)->setValue($password . Key::ENTER);
```
Comment (skjnldsv) [inline tests/acceptance/features/bootstrap/FilesAppSharingContext.php:431]: Love this

File: tests/acceptance/run.sh [hunk 2]
```diff
@@ -108,12 +108,12 @@ function prepareSelenium() {
 	SELENIUM_CONTAINER=selenium-nextcloud-local-test-acceptance
 
 	echo "Starting Selenium server"
-	docker run --detach --name=$SELENIUM_CONTAINER --publish 4444:4444 --publish 5900:5900 $DOCKER_OPTIONS selenium/standalone-firefox-debug:2.53.1-beryllium
+	docker run --detach --name=$SELENIUM_CONTAINER --publish 4444:4444 --publish 5900:5900 --shm-size 256m $DOCKER_OPTIONS selenium/standalone-chrome-debug:3.141.59
```
Comment (PVince81) [inline tests/acceptance/run.sh:111]: not sure if you tried this before: https://stackoverflow.com/a/41640458

if it didn't work, there's a note that says that the containers need to be marked as trusted in the repo settings

Comment (skjnldsv) [reply tests/acceptance/run.sh:111]: @PVince81 this file is also used to run selenium locally. So adding a drone-only config would not be compatible :)

**Comment by skjnldsv** [comment]
@danxuliu some tests crashes



```
And I log in with user unknownUser and password 123456acb         # LoginPageContext::iLogInWithUserAndPassword()
130 | unknown error: session deleted because of page crash
131 | from tab crashed
132 | (Session info: chrome=89.0.4389.82)
133 | (Driver info: chromedriver=89.0.4389.23 (61b08ee2c50024bab004e48d2b1b083cdbdac579-refs/branch-heads/4389@{#294}),platform=Linux 4.15.0-124-generic x86_64) (WebDriver\Exception\UnknownError)
134 | Then I see that the current page is the Files app                 # FilesAppContext::iSeeThatTheCurrentPageIsTheFilesApp()
135 | │
136 | ╳  Could not close connection (Behat\Mink\Exception\DriverException)
137 | │
138 | └─ @AfterScenario # ActorContext::cleanUpSessions()

```

File: tests/acceptance/run.sh (hunk 2 above)
Comment (PVince81) [reply tests/acceptance/run.sh:111]: ah, missing context. this was an answer to a chat message of @danxuliu where he said if I had an idea how to add it to Drone config. I just picked this line to answer as it was closest :-D

**Comment by danxuliu** [comment]
> @danxuliu some tests crashes

I know, please check the **Pending:** section in the pull request message :-P That is why it still has the _Developing_ tag ;-)

File: tests/acceptance/run.sh (hunk 2 above)
Comment (danxuliu) [reply tests/acceptance/run.sh:111]: > not sure if you tried this before: https://stackoverflow.com/a/41640458
>
> if it didn't work, there's a note that says that the containers need to be marked as trusted in the repo settings

As mentioned in the pull request description I have not tried it yet, but that does not seem to work in latest Drone even if the repository is trusted :-( But I still need to dig into all this ;-)

**Comment by skjnldsv** [comment]
> I know, please check the **Pending:** section in the pull request message :-P That is why it still has the _Developing_ tag ;-)

Well, it was pending for another pr during your vacations. Which I rebased and merged. So I assumed this one was maybe done :shrug:

**Comment by skjnldsv** [comment]
> **Pending:** shm size needs to be increased when run in Drone too. In theory [it should be enough to use `shm_size`](https://stackoverflow.com/questions/41546310/setting-up-drone-agents-to-have-custom-dev-shm/41640458#41640458), but [it seems that parameter no longer works](https://discourse.drone.io/t/shm-size-setting-not-respected-in-drone-1-0/4921)

Maybe move the tests to github actions?
Would be much faster

**Comment by danxuliu** [comment]
> Well, it was pending for another pr during your vacations. Which I rebased and merged. So I assumed this one was maybe done :shrug:

Fair enough ;-)

> skjnldsv marked this pull request as draft

:facepalm: I knew I forgot something xD

> Maybe move the tests to github actions?
> Would be much faster

I have no idea about GitHub actions :-( But I will take a look, thanks for the suggestion ;-)

Nevertheless, with the current Drone workers the tests are already quite fast. Those that take longer (the sharing ones) are mostly idling in some overcautious waits. I have took a look already at improving them, but failed miserably :-P They are a bit tricky ;-) But I hope that it will be eventually solved :-)

**Comment by skjnldsv** [comment]
>  I knew I forgot something xD

You didn't ;)
image

**Comment by PVince81** [comment]
seems it might be possible to disable shm, see https://developers.google.com/web/tools/puppeteer/troubleshooting#tips

that article points at issues with docker

**Comment by skjnldsv** [comment]
It worked! :rocket: 
Only the failing master test is still broken, but we can do this in a followup?

**Comment by danxuliu** [comment]
> seems it might be possible to disable shm, see https://developers.google.com/web/tools/puppeteer/troubleshooting#tips
> 
> that article points at issues with docker

Great finding, thanks! :-)

> It worked! :rocket:

Yay! :-D

> Only the failing master test is still broken, but we can do this in a followup?

No, luckily the failure from master is fixed now :-) The problem is that, even if Chrome uses the old protocol, in some cases it generates `element not interactable` errors, which [are from the new protocol](https://w3c.github.io/webdriver/#dfn-element-not-interactable).

However... I have been running tests literally for hours and the error was never thrown to really investigate it :cry: So I just added a tentative fix for it, but... I do not really know if the tests will always work or randomly fail :shrug:

**Comment by skjnldsv** [comment]
`continuous-integration/drone/pr — Build is passing` :tada: :rocket:

**Comment by skjnldsv** [comment]
LGTM

**Comment by danxuliu** [comment]
> `continuous-integration/drone/pr — Build is passing` :tada: :rocket:

The logs also showed some retries on `element not interactable` messages; that does not guarantee that the fix worked, as that message seems to be shown now also on `ElementNotVisible` exceptions (:shrug:). But it is likely that it worked, so... I would say merge this and keep an eye on failed tests in the next days ;-)

**Comment by skjnldsv** [comment]
A bit more reviewers @danxuliu ? :)

**Review by PVince81 — APPROVED** [review]
:+1:

**Comment by skjnldsv** [comment]
Unrelated failure of sqlite

**Comment by danxuliu** [comment]
All acceptance tests currently fail in Drone in pull requests that were branched before this pull request was merged.

It seems that there is a strange mismatch between the Selenium Docker image used and the expected one. I guess it is just some Drone caching issue.

**Comment by danxuliu** [comment]
> All acceptance tests currently fail in Drone in pull requests that were branched before this pull request was merged.
> 
> It seems that there is a strange mismatch between the Selenium Docker image used and the expected one. I guess it is just some Drone caching issue.

As mentioned it was probably some caching issue, as a pull request rebased after the merge got [green CI](https://drone.nextcloud.com/nextcloud/server/4446) (such a beautiful colour :-P ).

**Pull request merged**